portfolio/src/main/java/com/google/sps/servlets/ChineseDishesDataServlet.java

```diff
@@ -0,0 +1,96 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.FetchOptions;
+import java.io.IOException;
+import com.google.gson.Gson;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+/** Servlet that returns hashmap json of dishes and their votes */
+@WebServlet("/dishes-data")
+public class ChineseDishesDataServlet extends HttpServlet {
+    
+    String DISHES_ENTITY = "Dish";
+    String DISH_NAME_PROPERTY = "DishName";
+    String VOTE_AMOUNT_PROPERY = "VoteCnt";
+    String INDEX_URL = "/index.html";
+    String SELECT_INPUT_PARAM = "chinese-dishes-select";
+    String JSON_CONTENT = "application/json;";
+
+    /* Returns json of stored dish and votes hashmap */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        
+        Query query = new Query(DISHES_ENTITY);        
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        Map<String, Long> dishVotes = new HashMap<>();
+        for (Entity dish : results.asIterable()) {
+            String dishName = (String) dish.getProperty(DISH_NAME_PROPERTY);
+            long voteAmount = (long) dish.getProperty(VOTE_AMOUNT_PROPERY);
+            dishVotes.put(dishName, voteAmount);
+        }
+
+        Gson gson = new Gson();
+        String dishJson = gson.toJson(dishVotes);
+        response.setContentType(JSON_CONTENT);
+        response.getWriter().println(dishJson);
+    }
+
+    /* Updates dish vote amount at the datastore dish entity */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        long dishVoteCnt = 0;
+        Entity dishEntity;
+
+        String dishName = request.getParameter(SELECT_INPUT_PARAM);
+        Filter dishNameFilter = new Query.FilterPredicate(DISH_NAME_PROPERTY, FilterOperator.EQUAL, dishName);
+        Query dishQuery = new Query(DISHES_ENTITY).setFilter(dishNameFilter);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        FetchOptions fetchOption = FetchOptions.Builder.withDefaults();
+        List<Entity> dishResult = datastore.prepare(dishQuery).asList(fetchOption);
+
+        /* The dish doesn't exist in the DB -> creates new entity */
+        if (dishResult.isEmpty()) {
+            dishEntity = new Entity(DISHES_ENTITY);
+            dishEntity.setProperty(DISH_NAME_PROPERTY, dishName);
+        }
```

Comment:
no for action, just thinking material. 
What happens if 2 differnt uses are voting in the same time?